packages/firestore/src/api/database.ts

```diff
@@ -1618,6 +1645,45 @@ export class QuerySnapshot implements firestore.QuerySnapshot {
     return this._cachedChanges;
   }
 
+  public isEqual(other: firestore.QuerySnapshot): boolean {
+    if (!(other instanceof QuerySnapshot)) {
+      throw invalidClassError('isEqual', 'QuerySnapshot', 1, other);
+    }
+
+    if (
+      this.firestore !== other.firestore ||
+      !this._originalQuery.isEqual(other._originalQuery)
+    ) {
+      return false;
+    }
+    let snapshot: ViewSnapshsot = this._snapshot;
+    let otherSnapshot: ViewSnapshot = other._snapshot;
+    if (
+      !snapshot.query.isEqual(otherSnapshot.query) ||
+      !snapshot.docs.isEqual(otherSnapshot.docs) ||
```

Comment:
I'm kind of scared of this as this comparison is super expensive (it materializes the docs array). Unless we add `readTime`, I don't have a better solution to offer.

Reply:
Let me add a comment to warn user on the expensiveness.

Reply:
I don't think users should use QuerySnapshot.isEqual() very often (if ever), so I'm okay with it being slow.  We could add a TODO to improve the performance, but I think DocumentSnapshot.isEqual() is much more important (e.g. so devs can compare documents from two different QuerySnapshots to see what changed).